File: artifacts/libzahl-1.0/decompose/src/zsave.c
size_t zsave(z_t a, void *buffer)
{
  if (buffer)
  {
    char *buf = buffer;
    unsigned int buf_idx = 0;
    *((int *) buf) = a->sign, buf_idx += sizeof(int);
    *((size_t *) buf) = a->used, buf_idx += sizeof(size_t);
    if (!zzero(a))
    {
      memcpy(buf, a->chars, a->used * (sizeof(zahl_char_t)));
    }
  }
  return ((sizeof(int)) + (sizeof(size_t))) + ((zzero(a)) ? (0) : (a->used * (sizeof(zahl_char_t))));
}


/*** DEPENDENCIES:
inline static int zzero(z_t a)
{
  return !a->sign;
}


----------------------------
typedef struct 
{
  int sign;
  size_t used;
  size_t alloced;
  zahl_char_t *chars;
} z_t[1]
----------------------------
***/

File: artifacts/libzahl-1.0/manual_translation/src/zsave.rs
use crate::*;
use std::ptr;


pub fn zsave(a: &z_t, buffer: Option<&mut [u8]>) -> usize {
    let required_size = std::mem::size_of::<i32>() + std::mem::size_of::<usize>() + 
                       if zzero(a) { 0 } else { a[0].used * std::mem::size_of::<zahl_char_t>() };

    if let Some(buf) = buffer {
        if buf.len() < required_size {
            return required_size;
        }

        let mut buf_idx = 0;

        // Write sign
        let sign_slice = &mut buf[buf_idx..buf_idx + std::mem::size_of::<i32>()];
        sign_slice.copy_from_slice(&a[0].sign.to_ne_bytes());
        buf_idx += std::mem::size_of::<i32>();

        // Write used
        let used_slice = &mut buf[buf_idx..buf_idx + std::mem::size_of::<usize>()];
        used_slice.copy_from_slice(&a[0].used.to_ne_bytes());
        buf_idx += std::mem::size_of::<usize>();

        // Write chars if not zero
        if !zzero(a) {
            if let Some(chars) = &a[0].chars {
                let char_bytes = unsafe {
                    std::slice::from_raw_parts(
                        chars.as_ptr() as *const u8,
                        a[0].used * std::mem::size_of::<zahl_char_t>()
                    )
                };
                buf[buf_idx..buf_idx + char_bytes.len()].copy_from_slice(char_bytes);
            }
        }
    }

    required_size
}
